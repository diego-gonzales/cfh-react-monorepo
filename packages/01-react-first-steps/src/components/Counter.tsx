import { useState } from 'react';

const Counter = ({ value }: { value: number }) => {
  console.log('render');
  const [counter, setCounter] = useState(value);

  const handleAdd = () => {
    setCounter((c) => c + 1);
  };

  const handleSubstract = () => {
    setCounter(counter - 1);
  };

  const handleReset = () => setCounter(value);

  return (
    <div>
      <h2>Counter</h2>
      <p data-testid="test-counter">{counter}</p>

      <button aria-label="btn-add" onClick={handleSubstract}>
        -
      </button>
      <button aria-label="btn-reset" onClick={handleReset}>
        Reset
      </button>
      <button aria-label="btn-substract" onClick={handleAdd}>
        +
      </button>
    </div>
  );
};

export default Counter;
