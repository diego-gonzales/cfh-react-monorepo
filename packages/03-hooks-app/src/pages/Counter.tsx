import { useState } from 'react'
import { Button } from '../components/Button'

export const Counter = () => {
  const [myState, setMyState] = useState({
    counter1: 10,
    counter2: 20,
    counter3: 30,
  })

  const handleClick = () => {
    setMyState((prevState) => {
      return { ...prevState, counter1: prevState.counter1 + 1 }
    })
  }

  return (
    <>
      <h1>Counter 1: {myState.counter1}</h1>
      <h1>Counter 2: {myState.counter2}</h1>
      <h1>Counter 3: {myState.counter3}</h1>

      <Button onClick={handleClick}>+1</Button>
    </>
  )
}
