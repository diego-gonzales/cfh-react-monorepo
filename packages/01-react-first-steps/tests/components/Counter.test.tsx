import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import Counter from '../../src/components/Counter';

describe('Prueba en Counter component', () => {
  const initialValue = 100;

  test('should match with snapshot', () => {
    const { container } = render(<Counter value={100} />);

    expect(container).toMatchSnapshot();
  });

  test('should show the value passed as prop', () => {
    render(<Counter value={initialValue} />);

    expect(screen.getByText(initialValue)).toBeTruthy();
  });

  test('should increment the counter with button +', () => {
    render(<Counter value={initialValue} />);
    fireEvent.click(screen.getByText('+'));

    expect(screen.getByText(initialValue + 1)).toBeTruthy();
  });

  test('should decrement the counter with button -', () => {
    render(<Counter value={initialValue} />);
    fireEvent.click(screen.getByText('-'));
    // screen.debug();
    expect(screen.getByText(initialValue - 1)).toBeTruthy();
  });

  test('should reset the counter with button reset', () => {
    render(<Counter value={initialValue} />);
    // Otra forma de traer un elemento button por su nombre
    fireEvent.click(screen.getByRole('button', { name: 'btn-add' }));
    fireEvent.click(screen.getByRole('button', { name: 'btn-add' }));
    fireEvent.click(screen.getByRole('button', { name: 'btn-substract' }));
    fireEvent.click(screen.getByRole('button', { name: 'btn-reset' }));
    // screen.debug();
    expect(screen.getByTestId('test-counter').innerHTML).toContain(
      initialValue.toString()
    );
  });
});
