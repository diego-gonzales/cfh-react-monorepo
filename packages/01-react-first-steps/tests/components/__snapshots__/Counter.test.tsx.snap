// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en Counter component should match with snapshot 1`] = `
<div>
  <div>
    <h2>
      Counter
    </h2>
    <p
      data-testid="test-counter"
    >
      100
    </p>
    <button
      aria-label="btn-add"
    >
      -
    </button>
    <button
      aria-label="btn-reset"
    >
      Reset
    </button>
    <button
      aria-label="btn-substract"
    >
      +
    </button>
  </div>
</div>
`;
